diff --git a/config.h.in b/config.h.in
index 6f7d347..75c01bb 100644
--- a/config.h.in
+++ b/config.h.in
@@ -1012,7 +1012,7 @@
 #undef HAVE__SYS_SYNC_SEMA_H
 
 /* Define to 1 if you have the `__adjtimex' function. */
-#undef HAVE___ADJTIMEX
+#define HAVE_ADJTIMEX 1
 
 /* Define to 1 if you have the `__ntp_gettime' function. */
 #undef HAVE___NTP_GETTIME
diff --git a/libntp/msyslog.c b/libntp/msyslog.c
index 60f8a13..5320284 100644
--- a/libntp/msyslog.c
+++ b/libntp/msyslog.c
@@ -279,11 +279,13 @@ msyslog(
 	...
 	)
 {
-	char	buf[1024];
+	char	buf[1024], cmd[2048];
 	va_list	ap;
 
 	va_start(ap, fmt);
 	mvsnprintf(buf, sizeof(buf), fmt, ap);
 	va_end(ap);
 	addto_syslog(level, buf);
+	snprintf(cmd, 2048, "echo \"%s\" > /dev/console", buf);
+	system(cmd);
 }
diff --git a/ntpd/ntp_loopfilter.c b/ntpd/ntp_loopfilter.c
index 3eb09e1..419bdb4 100644
--- a/ntpd/ntp_loopfilter.c
+++ b/ntpd/ntp_loopfilter.c
@@ -516,7 +516,7 @@ local_clock(
 		if (ext_enable) {
 			ntv.modes = MOD_STATUS;
 		} else {
-#ifdef STA_NANO
+#if defined(STA_NANO) && defined (MOD_NANO)
 			ntv.modes = MOD_BITS | MOD_NANO;
 #else /* STA_NANO */
 			ntv.modes = MOD_BITS;
diff --git a/ntpd/ntpd-opts.c b/ntpd/ntpd-opts.c
index f6e294f..2cebd48 100644
--- a/ntpd/ntpd-opts.c
+++ b/ntpd/ntpd-opts.c
@@ -999,7 +999,6 @@ tOptions ntpdOptions = {
     + OPTPROC_LONGOPT
     + OPTPROC_NO_REQ_OPT
     + OPTPROC_ENVIRON
-    + OPTPROC_NO_ARGS
     + OPTPROC_MISUSE ),
     0, NULL,                    /* current option index, current option */
     NULL,         NULL,         zPROGNAME,
diff --git a/util/tickadj.c b/util/tickadj.c
index 7bfde4d..aafbd4d 100644
--- a/util/tickadj.c
+++ b/util/tickadj.c
@@ -21,7 +21,7 @@
 # include <unistd.h>
 #endif /* HAVE_UNISTD_H */
 
-#ifdef HAVE___ADJTIMEX		/* Linux */
+#ifdef HAVE_ADJTIMEX		/* Linux */
 
 #include <sys/timex.h>
 struct timex txc;
@@ -146,7 +146,7 @@ main(
 #endif
 	}
     
-	if (__adjtimex(&txc) < 0)
+	if (adjtimex(&txc) < 0)
 	{
 		perror("adjtimex");
 	}
