diff --git a/config.h.in b/config.h.in
index f4ce4a7..7cdbd3e 100644
--- a/config.h.in
+++ b/config.h.in
@@ -455,7 +455,7 @@
 #endif
 
 /* Define to rpl_malloc if the replacement function should be used. */
-#undef malloc
+//#undef malloc
 
 /* Define to `int' if <sys/types.h> does not define. */
 #undef mode_t
@@ -467,7 +467,7 @@
 #undef pid_t
 
 /* Define to rpl_realloc if the replacement function should be used. */
-#undef realloc
+//#undef realloc
 
 /* Define to `unsigned int' if <sys/types.h> does not define. */
 #undef size_t
