diff --git a/Makefile b/Makefile
deleted file mode 100644
index 6bba86c..0000000
--- a/Makefile
+++ /dev/null
@@ -1,5 +0,0 @@
-all:
-	-@echo "Please use ./configure first.  Thank you."
-
-distclean:
-	make -f Makefile.in distclean
diff --git a/Makefile.in b/Makefile.in
index c61aa30..f02f31a 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -255,9 +255,6 @@ maintainer-clean: distclean
 distclean: clean zconf zconf.h.cmakein docs
 	rm -f Makefile zlib.pc configure.log
 	-@rm -f .DS_Store
-	-@printf 'all:\n\t-@echo "Please use ./configure first.  Thank you."\n' > Makefile
-	-@printf '\ndistclean:\n\tmake -f Makefile.in distclean\n' >> Makefile
-	-@touch -r Makefile.in Makefile
 
 tags:
 	etags *.[ch]
